core/meterer/dynamodb_metering_store.go

```diff
@@ -256,30 +256,50 @@ func (s *DynamoDBMeteringStore) RollbackOnDemandPayment(ctx context.Context, acc
-	// Fetch the 3 bins start from the current bin
-	queryInput := &dynamodb.QueryInput{
-		TableName:              aws.String(s.reservationTableName),
-		KeyConditionExpression: aws.String("AccountID = :account AND ReservationPeriod >= :reservationPeriod"),
-		ExpressionAttributeValues: commondynamodb.ExpressionValues{
-			":account":           &types.AttributeValueMemberS{Value: accountID.Hex()},
-			":reservationPeriod": &types.AttributeValueMemberN{Value: strconv.FormatUint(reservationPeriod, 10)},
-		},
-		ScanIndexForward: aws.Bool(true),
-		Limit:            aws.Int32(MinNumBins),
-	}
-	bins, err := s.dynamoClient.QueryWithInput(ctx, queryInput)
+// GetPeriodRecords retrieves period records for multiple quorums efficiently.
+// This function is optimized for retrieving period records for all quorums in a single database operation.
+// The records start from the first reservation period for each quorum to at most numBins in length into the future.
+// quorumNumbers and reservationPeriods must have the same length and the same ordering.
+// Returns an array of PeriodRecords up to numBins in length, with records for each requested quorum.
+func (s *DynamoDBMeteringStore) GetPeriodRecords(
+	ctx context.Context,
+	accountID gethcommon.Address,
+	quorumNumbers []core.QuorumID,
+	reservationPeriods []uint64,
+	numBins uint32,
+) (map[core.QuorumID]*pb.PeriodRecords, error) {
+	if len(quorumNumbers) == 0 || len(reservationPeriods) != len(quorumNumbers) {
+		return nil, nil
+	}
+
+	// Prepare all keys for batch get
+	var keys []map[string]types.AttributeValue
+	for i, quorum := range quorumNumbers {
+		accountIDAndQuorum := accountID.Hex() + ":" + strconv.FormatUint(uint64(quorum), 10)
+		for j := 0; j < int(numBins); j++ {
+			key := map[string]types.AttributeValue{
+				"AccountID":         &types.AttributeValueMemberS{Value: accountIDAndQuorum},
+				"ReservationPeriod": &types.AttributeValueMemberN{Value: strconv.FormatUint(reservationPeriods[i]+uint64(j), 10)},
+			}
+			keys = append(keys, key)
+		}
+	}
+
+	items, err := s.dynamoClient.GetItems(ctx, s.reservationTableName, keys, true)
 	if err != nil {
-		return [MinNumBins]*pb.PeriodRecord{}, fmt.Errorf("failed to query payments for account: %w", err)
+		return nil, fmt.Errorf("failed to batch get period records for account: %w", err)
 	}
 
-	records := [MinNumBins]*pb.PeriodRecord{}
-	for i := 0; i < len(bins) && i < int(MinNumBins); i++ {
-		periodRecord, err := parsePeriodRecord(bins[i])
+	records := make(map[core.QuorumID]*pb.PeriodRecords)
+	for _, item := range items {
+		quorumNumber, periodRecord, err := parsePeriodRecord(item)
 		if err != nil {
-			return [MinNumBins]*pb.PeriodRecord{}, fmt.Errorf("failed to parse bin %d record: %w", i, err)
+			s.logger.Debug("Failed to parse period record", "err", err)
+			continue
+		}
+		records[quorumNumber] = &pb.PeriodRecords{
+			Records: []*pb.PeriodRecord{periodRecord},
```

Comment:
The comment for the function says `// Returns an array of PeriodRecords up to numBins in length, with records for each requested quorum.
` but seems we only return a size one array?

Reply:
ah nice catch! fixed in [commit](https://github.com/Layr-Labs/eigenda/pull/1620/commits/ddb8256f4ec865ffe4bd9fa44ae8a8eb52151372) and added some tests